src/crypto/hash.h

```diff
@@ -4,23 +4,128 @@
 #include "ds/buffer.h"
 #include "ds/json.h"
 
+#include <mbedtls/md.h>
+#include <mbedtls/pk.h>
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+
 #define FMT_HEADER_ONLY
 #include <fmt/format.h>
 #include <msgpack/msgpack.hpp>
 #include <ostream>
 
 namespace crypto
 {
+  enum class MDType
+  {
+    NONE = 0,
+    SHA1,
+    SHA256,
+    SHA384,
+    SHA512
+  };
+
+  using HashBytes = std::vector<uint8_t>;
+
+  class HashProviderBase
+  {
+  public:
+    virtual HashBytes Hash(const uint8_t*, size_t, MDType) const = 0;
+  };
+
+  class MBedHashProvider : public HashProviderBase
+  {
+  public:
+    static inline mbedtls_md_type_t get_md_type(MDType type)
+    {
+      switch (type)
+      {
+        case MDType::NONE:
+          return MBEDTLS_MD_NONE;
+        case MDType::SHA1:
+          return MBEDTLS_MD_SHA1;
+        case MDType::SHA256:
+          return MBEDTLS_MD_SHA256;
+        case MDType::SHA384:
+          return MBEDTLS_MD_SHA384;
+        case MDType::SHA512:
+          return MBEDTLS_MD_SHA512;
+        default:
+          throw std::runtime_error("Unsupported hash algorithm");
+      }
+      return MBEDTLS_MD_NONE;
+    }
+
+    virtual HashBytes Hash(const uint8_t* data, size_t size, MDType type) const
+    {
+      HashBytes r;
+      const auto mbedtls_md_type = get_md_type(type);
+      const auto info = mbedtls_md_info_from_type(mbedtls_md_type);
+      const auto hash_size = mbedtls_md_get_size(info);
+
+      r.resize(hash_size);
+
+      if (mbedtls_md(info, data, size, r.data()) != 0)
+        r.clear();
+
+      return r;
+    }
+  };
+
+  class OpenSSLHashProvider : public HashProviderBase
+  {
+  public:
+    static inline const EVP_MD* get_md_type(MDType type)
+    {
+      switch (type)
+      {
+        case MDType::NONE:
+          return nullptr;
+        case MDType::SHA1:
+          return EVP_sha1();
+        case MDType::SHA256:
+          return EVP_sha256();
+        case MDType::SHA384:
+          return EVP_sha384();
+        case MDType::SHA512:
+          return EVP_sha512();
+        default:
+          throw std::runtime_error("Unsupported hash algorithm");
+      }
+      return nullptr;
+    }
+
+    virtual HashBytes Hash(const uint8_t* data, size_t size, MDType type) const
+    {
+      HashBytes r;
+      unsigned int len = 0;
+      auto o_md_type = get_md_type(type);
+
+      r.resize(EVP_MD_size(o_md_type));
```

Comment:
I couldn't work out from a quick search if this was guaranteed to make a single heap allocation of the right size, or if line 100 would allocate a default, and then attempt to realloc on 104. I'd be tempted to create r only on line 104 with the right size straight away, just in case it's the latter, but I'm hopefully needlessly paranoid.

Reply:
Good observation - I'll move the allocation up so it's guaranteed only once.